vm/actor/src/builtin/miner/deadline_state.rs

```diff
@@ -0,0 +1,998 @@
+            // Nothing to do.
+            return Ok((BitField::new(), BitField::new(), PowerPair::zero()));
+        }
+
+        if let Some(partition_idx) = to_remove_set.iter().find(|&&i| i as u64 >= partition_count) {
+            return Err(
+                actor_error!(ErrIllegalArgument; "partition index {} out of range [0, {})", partition_idx, partition_count).into()
+            );
+        }
+
+        // Should already be checked earlier, but we might as well check again.
+        if !self.early_terminations.is_empty() {
+            return Err("cannot remove partitions from deadline with early terminations".into());
+        }
+
+        let mut new_partitions = Amt::<Partition, BS>::new(store);
+        let mut all_dead_sectors = Vec::<BitField>::with_capacity(to_remove_set.len());
+        let mut all_live_sectors = Vec::<BitField>::with_capacity(to_remove_set.len());
+        let mut removed_power = PowerPair::zero();
+
+        // TODO: maybe only unmarshal the partition if `to_remove_set` contains the
+        // corresponding index, like the Go impl does
+
+        old_partitions
+            .for_each(|partition_idx, partition| {
+                // If we're keeping the partition as-is, append it to the new partitions array.
+                if !to_remove_set.contains(&(partition_idx as usize)) {
+                    new_partitions.set(new_partitions.count(), partition.clone())?;
+                    return Ok(());
+                }
+
+                // Don't allow removing partitions with faulty sectors.
+                let has_no_faults = partition.faults.is_empty();
+                if !has_no_faults {
+                    return Err(actor_error!(ErrIllegalArgument; "cannot remove partition {}: has faults", partition_idx).into());
+                }
+
+                // Get the live sectors.
+                let live_sectors = partition.live_sectors();
+
+                all_dead_sectors.push(partition.terminated.clone());
+                all_live_sectors.push(live_sectors);
+                removed_power += &partition.live_power;
+
+                Ok(())
+            })
+            .map_err(|e| ActorError::downcast_wrap(e, "while removing partitions"))?;
+
+        self.partitions = new_partitions
+            .flush()
+            .map_err(|e| format!("failed to persist new partition table: {:?}", e))?;
+
+        let dead = BitField::union(&all_dead_sectors);
+        let live = BitField::union(&all_live_sectors);
+
+        // Update sector counts.
+        let removed_dead_sectors = dead.len() as u64;
+        let removed_live_sectors = live.len() as u64;
+
+        self.live_sectors -= removed_live_sectors;
+        self.total_sectors -= removed_live_sectors + removed_dead_sectors;
+
+        // Update expiration bitfields.
+        let mut expiration_epochs = BitFieldQueue::new(store, &self.expirations_epochs, quant)
+            .map_err(|e| format!("failed to load expiration queue: {:?}", e))?;
```

Comment:
here is another case of the previously mentioned. Nothing to change here, just noting